pkg/resource/subnet_group/custom_set_output.go

```diff
@@ -0,0 +1,95 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package subnet_group
+
+import (
+	"context"
+	"fmt"
+
+	svcsdk "github.com/aws/aws-sdk-go/service/memorydb"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	svcapitypes "github.com/aws-controllers-k8s/memorydb-controller/apis/v1alpha1"
+)
+
+const (
+	// The number of minutes worth of events to retrieve.
+	// 14 days in minutes
+	eventsDuration = 20160
+)
+
+func (rm *resourceManager) CustomDescribeSubnetGroupsSetOutput(
+	ctx context.Context,
+	r *resource,
+	resp *svcsdk.DescribeSubnetGroupsOutput,
+	ko *svcapitypes.SubnetGroup,
+) (*svcapitypes.SubnetGroup, error) {
+	if len(resp.SubnetGroups) == 0 {
+		return ko, nil
+	}
+	elem := resp.SubnetGroups[0]
+	err := rm.customSetOutputSupplementAPIs(ctx, r, elem, ko)
+	if err != nil {
+		return nil, err
+	}
+	fmt.Printf("DescribeSG %+v", ko.Status.Events)
+	return ko, nil
+}
+
+func (rm *resourceManager) customSetOutputSupplementAPIs(
+	ctx context.Context,
+	r *resource,
+	subnetGroup *svcsdk.SubnetGroup,
+	ko *svcapitypes.SubnetGroup,
+) error {
+	events, err := rm.provideEvents(ctx, r.ko.Spec.Name, 20)
+	if err != nil {
+		return err
+	}
+	ko.Status.Events = events
+	return nil
+}
+
+func (rm *resourceManager) provideEvents(
+	ctx context.Context,
+	subnetGroupName *string,
+	maxResults int64,
+) ([]*svcapitypes.Event, error) {
+	input := &svcsdk.DescribeEventsInput{}
+	input.SetSourceType("subnet-group")
+	input.SetSourceName(*subnetGroupName)
+	input.SetMaxResults(maxResults)
+	input.SetDuration(eventsDuration)
+	resp, err := rm.sdkapi.DescribeEventsWithContext(ctx, input)
+	rm.metrics.RecordAPICall("READ_MANY", "DescribeEvents-SubnetGroup", err)
+	if err != nil {
+		rm.log.V(1).Info("Error during DescribeEvents-SubnetGroup", "error", err)
+		return nil, err
+	}
+	events := []*svcapitypes.Event{}
+	if resp.Events != nil {
+		for _, respEvent := range resp.Events {
+			event := &svcapitypes.Event{}
+			if respEvent.Message != nil {
+				event.Message = respEvent.Message
+			}
+			if respEvent.Date != nil {
+				eventDate := metav1.NewTime(*respEvent.Date)
+				event.Date = &eventDate
+			}
+			events = append(events, event)
+		}
+	}
+	return events, nil
+}
```

Comment:
There's a lot of duplication here. Might be good to throw a `pkg/events/events.go` file together with a utility function like this:

```go
// NewDescribeEventsInput returns a populated DescribeEventsInput shape
func NewDescribeEventsInput(
	sourceName *string,
	sourceType string,
	maxResults int64,
) (*svcsdk.DescribeEventsInput {
	input := &svcsdk.DescribeEventsInput{}
	input.SetSourceType(sourceType)
	input.SetSourceName(*sourceName)
	input.SetMaxResults(maxResults)
	input.SetDuration(eventsDuration)
	return input
}

// EventsFromDescribe returns a slice of Event structs given the
// Output response shape from a DescribeEventsWithContext call
func EventsFromDescribe(
	resp *svcsdk.DescribeEventsOutput,
) ([]*svcapitypes.Event) {
	events := []*svcapitypes.Event{}
	if resp.Events != nil {
		for _, respEvent := range resp.Events {
			event := &svcapitypes.Event{}
			if respEvent.Message != nil {
				event.Message = respEvent.Message
			}
			if respEvent.Date != nil {
				eventDate := metav1.NewTime(*respEvent.Date)
				event.Date = &eventDate
			}
			events = append(events, event)
		}
	}
	return events
}
```

and then in all the `pkg/resource/$resource/*.go` files, you could reuse those functions, like so:

```go
import (
...
	"github.com/aws-controllers-k8s/memorydb-controller/pkg/events"
)
...
func (rm *resourceManager) provideEvents(
	ctx context.Context,
	subnetGroupName *string,
	maxResults int64,
) ([]*svcapitypes.Event, error) {
	input := events.NewDescribeEventsInput(subnetGroupName, "subnet-group", maxResults)
	resp, err := rm.sdkapi.DescribeEventsWithContext(ctx, input)
	rm.metrics.RecordAPICall("READ_MANY", "DescribeEvents", err)
	if err != nil {
		rm.log.V(1).Info("Error during DescribeEvents", "error", err)
		return nil, err
	}
	return events.EventsFromDescribe(resp), nil
}
```

That way you minimize the amount of duplicated code, reducing human copy/paste errors.

Reply:
@jaypipes Is this the first time we've seen custom code placed outside the resource folder?

Reply:
@RedbackThomson no. Elasticache has had a pkg/common for a while:

https://github.com/aws-controllers-k8s/elasticache-controller/blob/main/pkg/common/delta.go

Reply:
Created events.go to import and reuse.